### PR TITLE
Template part - add patterns placeholder selection.

### PR DESCRIPTION
**Pull request by Addison-Stavlo**
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->
Attempts to add the newly added (in #29602) patterns placeholder component to the template part block.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
First, verify the existing template part creation flow continues to work as expected.

Second, register some template part block patterns and test the creation flow when patterns are present.  We can register header and footer patterns in php using the example below.  For testing, I added this at the end of template-parts.php:

```
register_block_pattern(
	'template-part/header-test',
	array(
		'title'      => __( 'Header Test', 'gutenberg' ),
		'blockTypes' => array( 'core/template-part/header' ),
		'content'    => '<!-- wp:spacer {"height":70} -->
		<div style="height:70px" aria-hidden="true" class="wp-block-spacer"></div>
		<!-- /wp:spacer -->

		<!-- wp:columns {"align":"wide"} -->
		<div class="wp-block-columns alignwide"><!-- wp:column -->
		<div class="wp-block-column"><!-- wp:site-title /-->

		<!-- wp:site-tagline /--></div>
		<!-- /wp:column -->

		<!-- wp:column -->
		<div class="wp-block-column">
		<!-- wp:navigation {"orientation":"horizontal","itemsJustification":"right"} /-->
		</div>
		<!-- /wp:column --></div>
		<!-- /wp:columns -->

		<!-- wp:spacer {"height":70} -->
		<div style="height:70px" aria-hidden="true" class="wp-block-spacer"></div>
		<!-- /wp:spacer -->',
	)
);
register_block_pattern(
	'template-part/footer-test',
	array(
		'title'      => __( 'Footer Test', 'gutenberg' ),
		'blockTypes' => array( 'core/template-part/footer' ),
		'content'    => '<!-- wp:spacer {"height":70} -->
		<div style="height:70px" aria-hidden="true" class="wp-block-spacer"></div>
		<!-- /wp:spacer -->

		<!-- wp:columns {"align":"wide"} -->
		<div class="wp-block-columns alignwide"><!-- wp:column -->
		<div class="wp-block-column"><!-- wp:paragraph -->
		<p><a href="mailto:#">example@example.com<br></a>T. +00 (0)1 22 33 44 55</p>
		<!-- /wp:paragraph --></div>
		<!-- /wp:column -->
		<!-- wp:column -->
		<div class="wp-block-column"><!-- wp:paragraph {"align":"center"} -->
		<p class="has-text-align-center">2, Rue Louis-Boilly<br>Paris, France</p>
		<!-- /wp:paragraph --></div>
		<!-- /wp:column -->
		<!-- wp:column {"verticalAlignment":"center"} -->
		<div class="wp-block-column is-vertically-aligned-center"><!-- wp:social-links {"align":"right"} -->
		<ul class="wp-block-social-links alignright"><!-- wp:social-link {"url":"https://wordpress.org","service":"wordpress"} /-->
		<!-- wp:social-link {"url":"https://www.facebook.com/WordPress/","service":"facebook"} /-->
		<!-- wp:social-link {"url":"https://twitter.com/WordPress","service":"twitter"} /-->
		<!-- wp:social-link {"url":"https://www.youtube.com/wordpress","service":"youtube"} /--></ul>
		<!-- /wp:social-links --></div>
		<!-- /wp:column --></div>
		<!-- /wp:columns -->

		<!-- wp:spacer {"height":70} -->
		<div style="height:70px" aria-hidden="true" class="wp-block-spacer"></div>
		<!-- /wp:spacer -->

		<!-- wp:separator {"align":"wide","className":"is-style-twentytwentyone-separator-thick"} -->
		<hr class="wp-block-separator alignwide is-style-twentytwentyone-separator-thick"/>
		<!-- /wp:separator -->

		<!-- wp:columns {"align":"wide"} -->
		<div class="wp-block-columns alignwide"><!-- wp:column {"verticalAlignment":"center"} -->
		<div class="wp-block-column is-vertically-aligned-center"><!-- wp:site-title /--></div>
		<!-- /wp:column -->

		<!-- wp:column {"verticalAlignment":"center"} -->
		<div class="wp-block-column is-vertically-aligned-center"><!-- wp:paragraph {"align":"right"} -->
		<p class="has-text-align-right">Proudly powered by <a href="https://wordpress.org/">WordPress</a>.</p>
		<!-- /wp:paragraph --></div>
		<!-- /wp:column --></div>
		<!-- /wp:columns -->',
	)
);
```
Test that the patterns placeholder component appears for the Template Part block when we select "Create New" from the placeholder block flow for a header or footer ( block inserter -> "Header" or "Footer").

The "Header" and "Footer" variations should only show the header or footer pattern respectively.  The generic / "Template Part" variation should show no patterns selection step, since none are currently registered for the generic case.  We can test patterns to show up in the general case by adding `core/template-part` to their `blockTypes` array in registration.


## Screenshots <!-- if applicable -->

![Screen Shot 2021-04-27 at 3 23 51 PM](https://user-images.githubusercontent.com/28742426/116300406-dcf0b080-a76c-11eb-8437-72391b5fd65b.png)


## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

### CODE REVIEW

File: lib/full-site-editing/template-parts.php
Comment (Addison-Stavlo) [inline lib/full-site-editing/template-parts.php]: These are just copies of tt1-blocks template parts thrown in here as patterns for testing purposes.

A ❓ question to answer before shipping is - should we register any template-part block patterns in Gutenberg (or leave that to themes/plugins)?  And if so what should those patterns be?

**Review by Addison-Stavlo — COMMENTED** [review]
At this point the component works pretty well.  It appears when I would expect it to (creating a new 'blank' template part or removing all content from an existing one).

What is not working at this point is when we "choose" a specific pattern, it replaces the template part block itself instead of adding content inside of it.

File: lib/full-site-editing/template-parts.php
Comment (ntsekouras) [reply lib/full-site-editing/template-parts.php]: -cc @kjellr @mtias

Comment (mtias) [reply lib/full-site-editing/template-parts.php]: It largely depends on what comes up from the work on the navigation block and patterns for header areas. We would include some patterns for headers once those are ready.

File: packages/block-library/src/template-part/edit/index.js
Comment (ntsekouras) [inline packages/block-library/src/template-part/edit/index.js]: This needs to run only once.

```suggestion
	useEffect( () => {
		setStartingPattern( [] );
	}, [] );
```

**Comment by Addison-Stavlo** [comment]
some e2e's are expected to fail at this point - the patterns we have temporarily registered in this PR for testing purposes will trigger an extra selection step those tests don't account for. 

@jameskoster how does this flow feel currently?  Iirc from previous discussions the patterns selection should also be triggered if the user removes all their blocks from inside the template part? 🤔

**Comment by jameskoster** [comment]
> Iirc from previous discussions the patterns selection should also be triggered if the user removes all their blocks from inside the template part?

Yeah I posed that for discussion, but I don't think there was any strong feelings either way. If pattern selection were to move to a modal (https://github.com/WordPress/gutenberg/issues/31033) I suspect this flow wouldn't work so well, so we can probably discount it for now.

> The generic / "Template Part" variation should show both patterns in the placeholder. The "Header" and "Footer" variations should only show the header or footer pattern respectively.

This is working as described. The tricky thing here is determining which patterns to display for general template parts. 

In some ways it feels kind of strange to see Header / Footer patterns when I haven't specified an `area` for the template part yet. Especially as the area isn't subsequently set after selecting one of them.

Two options I can see:

1. Display _all_ "template part" patterns, and if the user selects a Header pattern, assign the header area automatically.
2. Only display template part patterns that have no specific area assigned to them.

---

Taking a step back, without knowing the intended area for the template part, the chances of us presenting a pattern that is relevant to the users intentions seems quite low, simply due to the enormous variety of template part use cases. This may need some more thought.

**Comment by Addison-Stavlo** [comment]
> Two options I can see:
> 1. Display all "template part" patterns, and if the user selects a Header pattern, assign the header area automatically.
> 2. Only display template part patterns that have no specific area assigned to them.

I think we could leave it up to who registers the pattern to determine where it shows up.  For example, I could register both of these example patterns so that they ONLY show up for their expected variations and never the general case.

So I registered the header pattern as `blockTypes: [ 'core/template-part', 'core/template-part/header' ]` (and similar for footer).  But you are right, neither of these patterns really make sense for the general/uncategorized case so we could choose not to show them there and just register them as `blockTypes: [ 'core/template-part/header' ]` (or footer).  

Leaving it up to how the pattern is registered would give flexibility, and if there ever IS a pattern that makes sense for the general case it can be registered to show up for the general case.  But as a standard 'rule' we could advise not registering header/footer patterns to show up for that case. 🤔

File: lib/full-site-editing/template-parts.php
Comment (kjellr) [reply lib/full-site-editing/template-parts.php]: This question is coming up for a lot of blocks. 😄 I've opened https://github.com/WordPress/gutenberg/issues/31231 for us to start exploring some of these patterns. 

It strikes me that when we combine template part patterns with all of our query block patterns, we're basically including a handful of mix-and-match themes for folks to choose from. 🤔

**Comment by Addison-Stavlo** [comment]
> I could register both of these example patterns so that they ONLY show up for their expected variations and never the general case.

Updated for this now.  So the general case should skip the patterns selection entirely now since we don't have any patterns registered that make sense for that case.

File: packages/block-library/src/template-part/edit/entity-content/patterns-setup.js
Comment (ntsekouras) [inline packages/block-library/src/template-part/edit/entity-content/patterns-setup.js]: Since we are using a custom pattern filtering callback we can omit this.

**Comment by Addison-Stavlo** [comment]
>  If pattern selection were to move to a modal (#31033) I suspect this flow wouldn't work so well, so we can probably discount it for now.

Awesome, that actually simplifies things quite a bit on the technical side by allowing us to use patterns selection to happen before entity creation.  Some version of patterns transforms would probably make more sense for the pre-existing template part case.

File: packages/block-library/src/template-part/edit/entity-content/patterns-setup.js
Comment (Addison-Stavlo) [reply packages/block-library/src/template-part/edit/entity-content/patterns-setup.js]: do you think we can omit clientId as well?  I guess its still assumed to be present in the selector so I guess not... but it feels sort of pointless in this circumstance since we are just getting patterns registered to `core/template-part/${variation}` anyways.

Comment (ntsekouras) [reply packages/block-library/src/template-part/edit/entity-content/patterns-setup.js]: >do you think we can omit clientId as well?

No because it checks for allowed blocks in the rootClientId that is fetched from the block's clientId

File: packages/block-library/src/template-part/edit/placeholder/index.js
Comment (ntsekouras) [inline packages/block-library/src/template-part/edit/placeholder/index.js]: I think we can use an `enum` like object for better readability here or plain strings.

File: packages/block-library/src/template-part/edit/placeholder/index.js
Comment (ntsekouras) [inline packages/block-library/src/template-part/edit/placeholder/index.js:20]: This must be a leftover 👍 . This was always empty array.

File: lib/full-site-editing/template-parts.php
Comment (ntsekouras) [inline lib/full-site-editing/template-parts.php]: Don't forget to remove these before merging. We can handle this in the related issue https://github.com/WordPress/gutenberg/issues/31231

**Review by ntsekouras — APPROVED** [review]
I'm very happy with this iteration. Great work Addison! 💯 

I've left a minor comment to address and I think a rebase will make the tests succeed, as they are related to the new components.

File: packages/block-library/src/template-part/edit/placeholder/index.js
Comment (Addison-Stavlo) [reply packages/block-library/src/template-part/edit/placeholder/index.js:20]: Yeah, I think its a leftover from the old implementation of the 'make template part' flow (selecting existing blocks => ellipsis menu) which has since been updated to make this unnecessary.

File: packages/block-library/src/template-part/edit/placeholder/index.js
Comment (Addison-Stavlo) [reply packages/block-library/src/template-part/edit/placeholder/index.js]: Sounds good!  Something along these lines? - 11c84b7

Comment (ntsekouras) [reply packages/block-library/src/template-part/edit/placeholder/index.js]: Yes, but I think it's better to have the props lowercase and you can use `numbers` as values. ex: `initial: 1`

**Pull request merged**
